docs/faq.rst

```diff
@@ -360,6 +360,62 @@ Yes, by setting :setting:`DNS_RESOLVER` to ``scrapy.resolver.CachingHostnameReso
 Note that by doing so, you lose the ability to set a specific timeout for DNS requests
 (the value of the :setting:`DNS_TIMEOUT` setting is ignored).
 
+.. _faq-dataclass-items:
+
+Can I use dataclasses as items?
+-------------------------------
+
+Support for :class:`dataclasses.dataclass` objects as items was added in version 2.1.
+This works natively in Python 3.7+, or using the `dataclasses backport`_ in Python 3.6.
+
+Most of the examples in this documentation assume you are using either :class:`dict`
+or :class:`~scrapy.item.Item` objects, and access their values in a dictionary-like
+manner. However, ``dataclass`` objects expose their values trough attributes instead,
+which means you might need to update your existing components
+(such as :ref:`item pipelines <topics-item-pipeline>`,
+:ref:`signal handlers <topics-signals>` handlers or
+:ref:`spider middlewares <topics-spider-middleware>`) to make them work correctly.
+
+Alternatively, you can use the
+:func:`~scrapy.utils.decorators.subscriptable_dataclass` decorator:
```

Comment:
I think we should think a bit more about the transition path, and update the docs accordingly. 

There are two possible ways people can adopt dataclasses.

Path 1: start using dataclasses. Realize that some middlewares in a project don't work. Use `@subscriptable_dataclass@dataclass` instead of `@dataclass`.

Path 2: start using dataclasses. Realize that some middlewares in a project don't work. Update the middlewares.

My feeling is that we're pushing users strongly to (1), because 

* There are no examples of how to write a middleware which is compatible with dataclasses, or how to update an existing middleware to work with dataclasses. 
* We don't have middleware examples in docs updated, so "by default" people are learning how to write non-compatible middlewares: "Most of the examples in this documentation assume you are using either :class:`dict`
or :class:`~scrapy.item.Item` objects";
* We don't mention that all built-in Scrapy components are updated to work with dataclasses (I think reader could get an impression they are not)
* We say "you might need to update your components ... [not sure how]. Alternatively, use a decorator [this is how]: ... This decorator is *also* useful when working with third-party components". 

I would prefer to push users to (2) more, i.e. to updating the middlewares, not using the decorator. The end result of (1) is that nobody can use dataclasses without a decorator, which means that Scrapy doesn't really work with third-party dataclasses. 

It will still be easier to throw a decorator to make it work, but we can reduce the gap significantly by providing helper functions and examples, promoting it as a good practice of writing middlewares. I think decorator should be presented as a "quick hack to make it work", and as a way to use a third-party component which is not updated yet, not as a default solution.

Thoughts? :)